### PR TITLE
Refactors credential code, moved to oauth2 subdir.

### PR DESCRIPTION
**Pull request by houglum**
This is part of the work to add additional credential support for the library (#579).  Next steps are to add support for implicit GCE, GKE, and GAE Flexible Environment credentials, along with adding retry logic (which may require refactoring the existing RawClient class in order to split pure HTTP functionality from Cloud-Storage-centric functionality... e.g. RawHttpClient, which RawStorageHttpClient could inherit from, etc., but that's another discussion and this set of changes is already quite large).

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1183)

### CODE REVIEW

**Comment by houglum** [comment]
Note sure what's up with the style checking in the TravisCI failure:

```
/v/google/cloud/storage/oauth2/credential_constants.cc:43:12: error: invalid case style for global constant 'kGoogleOAuthJwtGrantType' [readability-identifier-naming,-warnings-as-errors]
char const kGoogleOAuthJwtGrantType[] =
           ^~~~~~~~~~~~~~~~~~~~~~~~~
           K_GOOGLE_O_AUTH_JWT_GRANT_TYPE
```

That's... just flat-out wrong: https://google.github.io/styleguide/cppguide.html#Constant_Names

File: google/cloud/storage/oauth2/google_credentials.cc [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      return std::make_shared<AuthorizedUserCredentials<>>(contents);
+    }
+    if (cred_type == "service_account") {
+      return std::make_shared<ServiceAccountCredentials<>>(contents);
+    }
+    google::cloud::internal::RaiseRuntimeError(
+        "Unsupported credential type (" + cred_type +
+        ") when reading "
+        "Application Default Credentials file.");
+  } else {  // Check for implicit environment-based credentials.
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials.cc]: nit-ish: no `else` if we return or raise in the `if()` block then else is just adding indentation for no value.

File: google/cloud/storage/oauth2/google_credentials.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_OAUTH2_GOOGLE_CREDENTIALS_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_OAUTH2_GOOGLE_CREDENTIALS_H_
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/oauth2/anonymous_credentials.h"
+#include "google/cloud/storage/oauth2/authorized_user_credentials.h"
+#include "google/cloud/storage/oauth2/service_account_credentials.h"
+#include "google/cloud/storage/version.h"
+#include <memory>
+#include <sstream>
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials.h]: You are not using sstream in this header I think

File: google/cloud/storage/oauth2/google_credentials.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_OAUTH2_GOOGLE_CREDENTIALS_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_OAUTH2_GOOGLE_CREDENTIALS_H_
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/oauth2/anonymous_credentials.h"
+#include "google/cloud/storage/oauth2/authorized_user_credentials.h"
+#include "google/cloud/storage/oauth2/service_account_credentials.h"
+#include "google/cloud/storage/version.h"
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials.h]: Please avoid superfluous includes, inside Google these have no cost, we have massively parallel cached builds, outside Google these slow down the builds for our users.

File: google/cloud/storage/oauth2/google_credentials.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+/// Creates a service account credential.
+std::shared_ptr<ServiceAccountCredentials<>>
+    CreateServiceAccountCredentialsFromJsonFilePath(std::string);
+
+/// Creates a service account credential.
+std::shared_ptr<ServiceAccountCredentials<>>
+    CreateServiceAccountCredentialsFromJsonContents(std::string);
+
+// TODO(#579): Should we support loading service account credentials from a P12
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials.h]: We should break down that bug into smaller bugs, you are reusing it for P12, and GCE, and GKE, and GAE...

File: google/cloud/storage/oauth2/google_credentials.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+// file too? Other libraries do, but the JSON format is strongly preferred.
+
+namespace internal {
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials.h]: Consider moving this a file in the `internal/` subdirectory, I was wrong where I said (elsewhere) that it did not matter, it doesn't except that the Doxygen documentation will include this function :frown:

File: google/cloud/storage/oauth2/google_credentials.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Returns the path to the file containing Application Default Credentials, as
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials.h]: Consider:
```
/**
 * The name of the Application Default Credentials file if one is configured.
 *
 * Returns the ADC file configured by the GOOGLE_APPLICATION_CREDENTIALS, if set, or by
 * the default location, if a default location is defined, or the empty string.
 *
 * @see ... link for a full description from cloud.google.com
 */
```

The Doxygen documentation pulls the first line into the index / brief documentation. The detailed documentation is only used when users click on the specific function.

File: google/cloud/storage/oauth2/google_credentials_test.cc [hunk 7]
```diff
@@ -12,37 +12,69 @@
...
+using ::google::cloud::testing_util::EnvironmentVariableRestore;
 
 char const VAR_NAME[] = "GOOGLE_APPLICATION_CREDENTIALS";
 
-class CredentialsTest : public ::testing::Test {
+class GoogleCredentialsTest : public ::testing::Test {
+ public:
+  GoogleCredentialsTest()
+      : home_(internal::kGoogleAdcHomeVar),
+        override_variable_("GOOGLE_APPLICATION_CREDENTIALS") {}
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials_test.cc]: Shouldn't we have a constant for GOOGLE_APPLICATION_CREDENTIALS?

File: google/cloud/storage/oauth2/google_credentials_test.cc [hunk 8]
```diff
@@ -55,23 +87,37 @@ TEST_F(CredentialsTest, Insecure) {
  * requires valid keys and contacting Google's production servers, and would
  * make this an integration test.
  */
-TEST_F(CredentialsTest, LoadValidAuthorizedUserCredentials) {
+TEST_F(GoogleCredentialsTest, LoadValidAuthorizedUserCredentials) {
   char const filename[] = "authorized-user.json";
   std::ofstream os(filename);
-  os << R"""({
+  std::string contents_str = R"""({
```
Comment (coryan) [inline google/cloud/storage/oauth2/google_credentials_test.cc]: If you are doing this, move the definition of `os` closer to where it is first used .

File: google/cloud/storage/oauth2/refreshing_credentials.h [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace oauth2 {
+
+/**
+ * Base class for a credential type that must periodically refresh itself.
```
Comment (coryan) [inline google/cloud/storage/oauth2/refreshing_credentials.h]: That description suggest that this is a base class purely to reuse code.  Can we use composition?

File: google/cloud/storage/oauth2/service_account_credentials.h [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+#include "google/cloud/storage/internal/curl_request_builder.h"
+#include "google/cloud/storage/internal/http_response.h"
+#include "google/cloud/storage/internal/nljson.h"
+#include "google/cloud/storage/internal/openssl_util.h"
+#include "google/cloud/storage/oauth2/credential_constants.h"
+#include "google/cloud/storage/oauth2/refreshing_credentials.h"
+#include "google/cloud/storage/version.h"
+#include <chrono>
+#include <ctime>
+#include <set>
```
Comment (coryan) [inline google/cloud/storage/oauth2/service_account_credentials.h]: Is that used in this header?

File: google/cloud/storage/oauth2/service_account_credentials.h [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/internal/curl_request_builder.h"
+#include "google/cloud/storage/internal/http_response.h"
+#include "google/cloud/storage/internal/nljson.h"
+#include "google/cloud/storage/internal/openssl_util.h"
+#include "google/cloud/storage/oauth2/credential_constants.h"
+#include "google/cloud/storage/oauth2/refreshing_credentials.h"
+#include "google/cloud/storage/version.h"
+#include <chrono>
+#include <ctime>
```
Comment (coryan) [inline google/cloud/storage/oauth2/service_account_credentials.h]: Is that used in this header?

**Comment by coryan** [comment]
Should we close this PR and start a new one?

**Comment by houglum** [comment]
Yeah, definitely.  I intended on doing this earlier this week but got distracted :) Thanks for the reminder; closing.

**Pull request closed without merging**